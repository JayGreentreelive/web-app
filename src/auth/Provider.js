import React, {
    useReducer,
    useEffect,
    useState,
    useContext,
    createContext
} from 'react'

export const AUTH_TOKEN_KEY = 'auth-token'
const initialState = ''
const localState = localStorage.getItem(AUTH_TOKEN_KEY)
const AuthContext = createContext()
const useAuth = () => useContext(AuthContext)

let reducer = (token, newToken) => {
    if (newToken === null) {
        localStorage.removeItem(AUTH_TOKEN_KEY)
        return initialState
    }
    return newToken || token
}

const AuthProvider = (props) => {
    const [token, setToken] = useReducer(reducer, localState || initialState)
    const [isLoggedIn, setIsLoggedIn] = useState(!!token)
    const logout = () => {
        setToken(null)
    }

    useEffect(() => {
        localStorage.setItem(AUTH_TOKEN_KEY, token)
        setIsLoggedIn(!!token)
    }, [token])

    return (
        <AuthContext.Provider
            value={{
                token,
                setToken,
                isLoggedIn,
                logout
            }}
        >
            {props.children}
        </AuthContext.Provider>
    )
}

export {
    AuthContext,
    AuthProvider,
    useAuth
}
