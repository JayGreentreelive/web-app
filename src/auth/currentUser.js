import { useState, useEffect } from 'react'
import { useLazyQuery } from 'react-apollo'
import { get } from 'lodash'
import { useAuth } from './'
import gql from 'graphql-tag'
import moment from 'moment'

const GET_CURRENT_PERSON = gql`
    query {
        currentUser {
            profile {
                firstName
                lastName 
            }
        }
    }
`

const useCurrentUser = (fields) => {
    const { token, isLoggedIn, logout } = useAuth()
    const [doRefetch, setRefetch] = useState(false)
    const [getCurrentUser, {
        data,
        loading,
        error,
        called,
        refetch
    }] = useLazyQuery(GET_CURRENT_PERSON)
    let currentUser = get(data, 'currentUser.profile', null)

    useEffect(() => {
        if (isLoggedIn) setRefetch(true)
    }, [token])

    try {
        if (doRefetch) {
            // If the state of doRefetch is true,
            //  attempt to refetch the data, then set the
            //  state to false to avoid an infinite loop
            refetch().catch(e => console.log({ e }))
            setRefetch(false)
            currentUser = null // data does not automatically update, so this needs to be set manually
        } else if (error && !doRefetch && !loading) {
            // Log error and log out if there is an error found
            //  Ignore the error if doRefetch is set to true because
            //  the error is left over from a previous request and
            //  ignore errors if the request is still loading
            console.error("Authentication error: logging out")
            logout()
            currentUser = null // data does not automatically update, so this needs to be set manually
        }

        // Request data on initial load:
        //  if the request has not already been called,
        //  check to see if the use is logged in before calling
        if (!called && isLoggedIn) {
            getCurrentUser()
        }
    } catch (e) {
        console.error('GQL error:', { e })
        currentUser = null // data does not automatically update, so this needs to be set manually
    }

    return {
        currentUser,
        loading,
        error
    }
}

export {
    useCurrentUser
}