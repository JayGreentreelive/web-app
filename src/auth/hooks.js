import React, { useState, useEffect } from 'react'

const LOGIN_KEY = 'login-status'
const AUTH_TOKEN_KEY = 'auth-token'

export const useLogin = () => {
    const [isLoggedIn, setIsLoggedIn] = useState(localStorage.getItem(LOGIN_KEY) || false)

    useEffect(() => {
        localStorage.setItem(LOGIN_KEY, isLoggedIn)
    }, [isLoggedIn])

    const setLoginStatus = (status) => setIsLoggedIn(status)

    return { isLoggedIn, setLoginStatus }
}

export const useAuth = () => {
    const [authToken, setAuthToken] = useState(localStorage.getItem(AUTH_TOKEN_KEY))
    const [isLoggedIn, setIsLoggedIn] = useState(authToken !== '')

    useEffect(() => {
        localStorage.setItem(AUTH_TOKEN_KEY, authToken)
        setIsLoggedIn(authToken !== '')
    }, [authToken, setIsLoggedIn])

    return {
        isLoggedIn,
        setAuthToken,
        authToken
    }
}