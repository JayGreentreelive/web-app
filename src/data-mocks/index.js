// Articles
import * as ArticleCategories from './article-categories'
import * as ArticleDetails from './article-details'
import * as RelatedArticles from './related-articles'
import * as TopThreeArticles from './top-three-articles'
import * as ArticleList from './article-list'

export const Articles = {
    ...ArticleCategories,
    ...ArticleDetails,
    ...RelatedArticles,
    ...TopThreeArticles,
    ...ArticleList,
}

// Features
import * as EmailCapture from './email-capture'

export const Features = {
    ...EmailCapture,
}

// Login
import * as LoginMocks from './login'

export const Login = {
    ...LoginMocks,
}

// Events
import * as EventConnected from './event-connected'
import * as EventListConnected from './event-list-connected'

export const Events = {
    ...EventConnected,
    ...EventListConnected
}

// Card Feed
import * as CardFeedMocks from './card-feed'

export const CardFeed = {
    ...CardFeedMocks,
}