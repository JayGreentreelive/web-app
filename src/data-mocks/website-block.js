import getWebPageBlockItems from '../queries/getWebPageBlockItems'

export const GET_BLOCK_ITEM_MOCK = {
    request: {
        query: getWebPageBlockItems,
        variables: { id: "WebsiteBlock" }
    },
    result: {
        data: {
            title: 'WebsiteBlock-Title'
        },
    },
}