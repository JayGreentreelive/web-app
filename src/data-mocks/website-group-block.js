import getGroupContentItems from '../queries/getGroupContentItems'
import GET_BLOCK_ITEM from '../data-mocks/website-block'

export const GET_GROUP_ITEM_MOCK = {
    request: {
        query: getGroupContentItems,
        variables: { id: "GroupBlock" }
    },
    result: {
        data: {
            node: {
                childContentItemsConnection: {
                    edges: [
                        { node: GET_BLOCK_ITEM }, 
                    ]
                }
            }
        },
    },
}