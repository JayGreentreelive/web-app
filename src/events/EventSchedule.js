import React, { useState } from 'react'
import { useQuery } from 'react-apollo'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {
  get,
  find
} from 'lodash'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
  faCalendarAlt,
  faClock,
  faAngleDown
} from "@fortawesome/fontawesome-pro-regular"
import {
  Dropdown
} from 'react-bootstrap'
import moment from 'moment'

import {
  Button,
  Card,
} from '../ui'
import Icon from './eventIcon'

import { useAuth } from '../auth'
import { CAMPUS_KEY } from '../keys'
import { getDirectionsUrl } from '../utils'
import { GET_EVENT_SCHEDULES } from './queries'

const row = classnames(
  'row',
  'my-4'
)

const ScheduleList = ({ dates }) =>
  dates.map((n, i) => {
    const date = moment(n)

    return (
      <div
        className={row}
        key={`EventSchedule:${i}`}
      >
        <div className="col-12">
          <h3>
            <Icon
              icon={faCalendarAlt}
              className="mr-2"
            />
            {date.format('ddd MMM D')}
          </h3>
        </div>

        <div
          className='col-12'
          key={i}
        >
          <h4
            className='font-weight-normal'
          >
            <Icon
              icon={faClock}
              className="mr-2"
            />
            {date.format('LT')}
          </h4>
        </div>
      </div>
    )
  })

const CampusSelectToggle = React.forwardRef(({ children, onClick }, ref) => (
  <a
    href=""
    className="h6 text-secondary font-weight-normal"
    ref={ref}
    onClick={e => {
      e.preventDefault()
      onClick(e)
    }}
  >
    {children}
    <span
      className="ml-2"
    >
      <FontAwesomeIcon
        icon={faAngleDown}
      />
    </span>
  </a>
))

const parseSchedulesByCampus = (schedules = []) => {
  // Set an empty array for adding dates
  let parsed = []

  // Loop through each schedule and add new campus specific dates
  schedules.forEach(schedule => {
    const { campuses } = schedule
    const campusDates = campuses.map(campus => ({
      campus,
      dates: schedule.dates,
      location: schedule.location
    }))

    parsed = [...parsed, ...campusDates]
  })

  return parsed
}

const EventSchedule = ({
  id,
  defaultCampus,
  callsToAction,
}) => {
  const [payload, setPayload] = useState({})
  const { isLoggedIn } = useAuth()
  const { loading, error, data } = useQuery(
    GET_EVENT_SCHEDULES,
    {
      variables: { id },
      onCompleted: (data) => {
        const schedules = get(data, 'node.childContentItemsConnection.edges', [])
          .map(edge => edge.node)
        const schedulesByCampus = parseSchedulesByCampus(schedules)
        const selectedCampus = schedulesByCampus.length == 1
          ? schedulesByCampus[0]
          : find(schedulesByCampus, (n) => n.campus.name === defaultCampus)

        setPayload(selectedCampus)
      }
    }
  )

  // Map the children to a collection of content items
  const schedules = get(data, 'node.childContentItemsConnection.edges', [])
    .map(edge => edge.node)

  const schedulesByCampus = parseSchedulesByCampus(schedules)

  return <Card
    className="py-3"
    loading={loading}
    error={get(error, 'message', null)}
  >
    <div>
      <div className="row">
        <div className="col">
          <h3 className="text-dark mb-0">
            {get(payload, 'campus.name', '')}
          </h3>
          <Dropdown
            onSelect={(key, e) => {
              e.preventDefault()
              const campusStr = key.replace('#/', '')

              setPayload(find(schedulesByCampus, (n) => n.campus.name === campusStr))
              if (!isLoggedIn) localStorage.setItem(CAMPUS_KEY, campusStr)
            }}
          >
            <Dropdown.Toggle
              variant="link"
              id="event-campus-selection"
              as={CampusSelectToggle}
            >
              {get(payload, 'campus.name', '') === ''
                ? 'Select Campus'
                : 'Change Campus'}
            </Dropdown.Toggle>

            <Dropdown.Menu>
              {schedulesByCampus.map((n, i) =>
                <Dropdown.Item
                  key={i}
                  href={`#/${n.campus.name}`}
                  active={n === get(payload, 'campus', '')}
                >
                  {n.campus.name}
                </Dropdown.Item>
              )}
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </div>
      <ScheduleList dates={get(payload, 'dates', [])} />

      <hr></hr>

      <div className="row mt-4">
        <div className="col">
          <a
            className="text-dark"
            href={getDirectionsUrl(get(payload, 'location', ''))}
            target="_blank"
          >
            {get(payload, 'location', '')}
          </a>
        </div>
      </div>

      {callsToAction.map((n, i) => (
        <div className="row my-4" key={i}>
          <div className="col">
            <a
              className={classnames(
                'btn',
                'btn-primary',
                'btn-block',
              )}
              href={n.action}
            >
              {n.call}
            </a>
          </div>
        </div>
      ))}
    </div>
  </Card>
}

EventSchedule.propTypes = {
  id: PropTypes.string.isRequired,
  defaultCampus: PropTypes.string,
  callsToAction: PropTypes.arrayOf(
    PropTypes.shape({
      call: PropTypes.string,
      action: PropTypes.string,
    })
  )
}

EventSchedule.defaultProps = {
  defaultCampus: '',
  callsToAction: []
}

export default EventSchedule
