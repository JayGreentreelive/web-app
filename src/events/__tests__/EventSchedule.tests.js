import React from 'react'
import { act, render } from '@testing-library/react'
import wait from 'waait'

import { AuthProvider } from '../../auth'
import EventSchedule from '../EventSchedule'
import { Events } from '../../data-mocks'

const { TEST_EVENT_JSON } = Events

let component = null

describe('EventSchedule', () => {
    it("renders the schedule of an event", async () => {
        component = render(
            <AuthProvider>
                <EventSchedule {...TEST_EVENT_JSON} />
            </AuthProvider>
        )

        const { container } = component
        expect(container).toMatchSnapshot()
    })
})