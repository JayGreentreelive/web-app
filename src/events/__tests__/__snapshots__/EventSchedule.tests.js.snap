// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventSchedule renders an empty schedule 1`] = `
<div>
  <div
    class="card border-0 shadow mb-3"
  >
    <div
      class="card-body"
    >
      <div
        class="py-3"
      >
        <h3
          class="mb-n4"
        >
          Get Started
        </h3>
        <div
          class="mt-5"
        >
          <a
            class="btn btn-primary btn-block my-3"
            href="/#"
            target=""
          >
            Register Here
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EventSchedule renders the schedule of an event 1`] = `
<div>
  <div
    class="card border-0 shadow mb-3"
  >
    <div
      class="card-body"
    >
      <div
        class="dropdown"
        id="event-campus-selection"
      >
        <div
          class="w-100"
          style="vertical-align: middle; cursor: pointer;"
        >
          <span
            class="h4"
          >
            Select Location
            <span
              class="ml-2 float-right"
              style="font-size: 22px;"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-angle-down fa-w-10 "
                data-icon="angle-down"
                data-prefix="far"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M151.5 347.8L3.5 201c-4.7-4.7-4.7-12.3 0-17l19.8-19.8c4.7-4.7 12.3-4.7 17 0L160 282.7l119.7-118.5c4.7-4.7 12.3-4.7 17 0l19.8 19.8c4.7 4.7 4.7 12.3 0 17l-148 146.8c-4.7 4.7-12.3 4.7-17 0z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="card border-0 shadow mb-3"
  >
    <div
      class="card-body"
    >
      <div
        class="py-3"
      >
        <div
          class=""
        />
      </div>
    </div>
  </div>
</div>
`;
